### ISSUE TITLE
Add PSR1.Files.SideEffects to WordPress-Extra?

### ISSUE THREAD

**Issue by GaryJones** [WordPress-Coding-Standards/WordPress-Coding-Standards#1069]
`[PSR1.Files.SideEffects](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/PSR1/Sniffs/Files/SideEffectsSniff.php) is the sniff that covers the following bit of [PSR-1](http://www.php-fig.org/psr/psr-1/#23-side-effects):

> A file SHOULD declare new symbols (classes, functions, constants, etc.) and cause no other side effects, or it SHOULD execute logic with side effects, but SHOULD NOT do both.
>
> The phrase “side effects” means execution of logic not directly related to declaring classes, functions, constants, etc., merely from including the file.
>
> “Side effects” include but are not limited to: generating output, explicit use of require or include, connecting to external services, modifying ini settings, emitting errors or exceptions, modifying global or static variables, reading from or writing to a file, and so on.

Most readers here will agree that keeping symbol definitions separate from files that actively do something is a good plan, particularly when it comes to autoloaded files.

The issue for not implementing this in `WordPres-Extra` is that there might be _many_ plugins and themes (let alone WP core) that do something like:

```php
add_action( '...', 'my_function' );
/**
 * DocBlock
 */
function my_function() {
    // ...
}
```

Refactoring that to move the `add_action()` / `add_filter()` call is not trivial, and can make non-OO code harder to follow.

@WordPress-Coding-Standards/wpcs-admins Thoughts?

**Comment by Rarst** [comment]
I think this is too far out from _average_ WP code to be more useful than annoying. It's also not _specific_ to writing _WordPress_ code in the least.

While I am huge fan of this rule personally, I don’t feel it’s in scope even for the Extra.

**Comment by JDGrimes** [comment]
FWIW, although core probably doesn't follow this rule fully, there was https://core.trac.wordpress.org/ticket/30947, which moved most of the `add_action()`/`add_filter()` calls out of the functions files and to their own file.

And I usually follow that pattern in my own plugins.

Of course, the default pattern currently employed by plugins and themes is to place the hook registration above or below the function definition. This is always going to be the format of code examples that people copy and paste too, because its easiest to just copy one block of code to one file.

So I think that this is really not a helpful rule for the average WP developer, unless they are making the choice as a personal preference. Unless there are strong arguments specifically relating to why actions and hooks shouldn't be inline that apply to the average plugin.

But maybe there could be a compromise position: allow actions and hooks, but not other code? That is a best practice that should generally be followed by all plugins; most code shouldn't run directly, it should be in a function hooked to some action or filter. (And yes, I see plugins violate this all the time, even doing crazy stuff like having an `add_option()` call right in the main file.) So a rule like that might be useful.

**Comment by Rarst** [comment]
> But maybe there could be a compromise position: allow actions and hooks, but not other code?

Rabbit hole. If we allow actions and hooks, next we need to allow _conditional_ actions and hooks. It just turns into allowing [nearly] arbitrary code, which defeats the point of the rule overall.

**Comment by jrfnl** [comment]
I generally have this sniff in my personal `phpcs.xml` file, but not in a project `phpcs.xml.dist`. It helps me minimize the occurrences, but rarely allows me to completely avoid it for WP project for all the reasons mentioned above.

So, no, I would not add it to WPCS for any of the ruleset.

What about adding a new wiki page for "Interesting other sniffs" ? There we could list sniffs which do not necessarily qualify for WPCS, but could be interesting to people anyhow.
We could also list some additional standards there which sniff for typical best practices which are too far removed from most WP projects, but could be of interest to the more advanced devs.

**Comment by GaryJones** [comment]
Thanks all. @Rarst's comment of "too far out from _average_ WP code" summed up nicely what my thoughts would be, but I wanted to have the discussion anyway (so we've got something to refer to).

I think a wiki page would be good.

**Comment by grappler** [comment]
> That is a best practice that should generally be followed by all plugins; most code shouldn't run directly, it should be in a function hooked to some action or filter.

I would love to see a sniff for this. I keep on seeing theme developers directly run `add_theme_support()` without it being hooked to an action. This would be something that could be added to theme sniffs.

**Comment by jrfnl** [comment]
@grappler I don't think that would be feasible.
You're basically asking for a sniff which checks that (nearly) everything is scoped, but then you'd run into the same issues as discussed in #1043 - template files being included from within a function, but the code in the template file itself is not scoped.

Also see @Rarst's comment: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/1069#issuecomment-318082759

**Comment by grappler** [comment]
It is not a big issue. I think we could solve the theme problem in different ways. Like only checking for specific functions that are commonly misused but are not used in templates.

Thinking aloud here. I wonder if it would be possible to have info in the template doc header that states that it is a template.

**Comment by GaryJones** [comment]
> Thinking aloud here. I wonder if it would be possible to have info in the template doc header that states that it is a template.

Part of a project's ruleset could exclude sniff(s) for a particular file path/name pattern.

(...unless you can think of more uses for having templates / view files having a header - in which case, this would need getting included in WP core, before we could do any sniff against it.)

**Comment by grappler** [comment]
> Part of a project's ruleset could exclude sniff(s) for a particular file path/name pattern.

I think a pattern will be difficult as in `functions.php` you could load a file `require get_template_directory() . '/custom-header.php';` or in  `index.php` load the same file using `get_template_part( 'custom', 'header' );`

> ...unless you can think of more uses for having templates / view files having a header - in which case, this would need getting included in WP core, before we could do any sniff against it.

I can't think of any at the moment. I suppose the easiest solution is just to exclude a few files from the sniffs.
